File: WebApiServer/Models/Basket.cs
namespace Models
{
    public class Basket
    {
        public int Id { get; set; }
        public required string BasketId { get; set; }
        public List<BasketItem> Items { get; set; } = [];

        public void AddItem(Product product, int quantity)
        {
            if (product == null)
            {
                throw new ArgumentNullException(nameof(product));
            }
            if (quantity <= 0)
            {
                throw new ArgumentOutOfRangeException("Quantity shoudl be greater than zero ", nameof(quantity));
            }
            var existingItem = FindItem(product.Id);
            if (existingItem == null)
            {
                Items.Add(new BasketItem
                {
                    ProductId = product.Id,
                    Quantity = quantity,
                    Product = product
                });
            }
            else
            {
                existingItem.Quantity += quantity;
            }
        }

        public void RemoveItem(int productId, int quantity)
        {
            if (quantity <= 0)
            {
                throw new ArgumentOutOfRangeException("Quantity shoudl be greater than zero ", nameof(quantity));
            }
            var item = FindItem(productId);
            if (item == null)
            {
                return;
            }
            item.Quantity -= quantity;
            if (item.Quantity <= 0)
            {
                Items.Remove(item);
            }
        }

        private BasketItem? FindItem(int productId)
        {
            return Items.FirstOrDefault(i => i.ProductId == productId);
        }
    }
}

File: client/src/app/models/basket.ts
import { Product } from "./Product";

export type Basket = {
    basketId: string
    items: Item[]
  }
  
  export class Item  {
  constructor(prduct: Product, quantity: number) {
    this.id = prduct.id;
    this.name = prduct.name;
    this.price = prduct.price;
    this.pictureUrl = prduct.pictureUrl;
    this.brand = prduct.brand;
    this.type = prduct.type;
    this.quantity = quantity;
  }

    id: number
    name: string
    price: number
    pictureUrl: string
    brand: string
    type: string
    quantity: number
  }
  